contracts/contracts/strategies/sonic/SonicSwapXAMOStrategy.sol

```diff
@@ -0,0 +1,738 @@
+            osToMint,
+            wsDepositAmount,
+            osDepositAmount,
+            lpTokens
+        );
+    }
+
+    /***************************************
+                Assets and Rewards
+    ****************************************/
+
+    /**
+     * @notice Get the wS value of assets in the strategy and SwapX pool.
+     * The value of the assets in the pool is calculated assuming the pool is balanced.
+     * This way the value can not be manipulated by changing the pool's token balances.
+     * @param _asset      Address of the Wrapped S (wS) token
+     * @return balance    Total value in wS.
+     */
+    function checkBalance(address _asset)
+        external
+        view
+        override
+        returns (uint256 balance)
+    {
+        require(_asset == ws, "Unsupported asset");
+
+        // wS balance needed here for the balance check that happens from vault during depositing.
+        balance = IERC20(ws).balanceOf(address(this));
+
+        // This assumes 1 gauge LP token = 1 pool LP token
+        uint256 lpTokens = IGauge(gauge).balanceOf(address(this));
+        if (lpTokens == 0) return balance;
+
+        // Add the strategy’s share of the wS and OS tokens in the SwapX pool if the pool was balanced.
+        balance += _lpValue(lpTokens);
+    }
+
+    /**
+     * @notice Returns bool indicating whether asset is supported by strategy
+     * @param _asset Address of the asset
+     */
+    function supportsAsset(address _asset) public view override returns (bool) {
+        return _asset == ws;
+    }
+
+    /**
+     * @notice Collect accumulated SWPx (and other) rewards and send to the Harvester.
+     */
+    function collectRewardTokens()
+        external
+        override
+        onlyHarvester
+        nonReentrant
+    {
+        // Collect SWPx rewards from the gauge
+        IGauge(gauge).getReward();
+
+        _collectRewardTokens();
+    }
+
+    /***************************************
+        Internal SwapX Pool and Gauge Functions
+    ****************************************/
+
+    /**
+     * @dev Calculate the required amount of OS to mint based on the wS amount.
+     * This ensures the proportion of OS tokens being added to the pool matches the proportion of wS tokens.
+     * For example, if the added wS tokens is 10% of existing wS tokens in the pool,
+     * then the OS tokens being added should also be 10% of the OS tokens in the pool.
+     * @param _wsAmount Amount of Wrapped S (wS) to be added to the pool.
+     * @return osAmount Amount of OS to be minted and added to the pool.
+     */
+    function _calcTokensToMint(uint256 _wsAmount)
+        internal
+        view
+        returns (uint256 osAmount)
+    {
+        (uint256 wsReserves, uint256 osReserves, ) = IPair(pool).getReserves();
+        require(wsReserves > 0, "Empty pool");
+
+        // OS to add = (wS being added * OS in pool) / wS in pool
+        osAmount = (_wsAmount * osReserves) / wsReserves;
```

Comment:
🟥 🟥 rekt, rekt.

Attacker front runs a deposit transaction by depositing a huge amount on either side. We deposit matching. When the pool is imbalanced, you get less LP per funds on deposit. (You get more on a withdraw). We lose money, get back less LP tokens. Attacker rebalances pool, has more LP tokens (since they profited from us getting less).

Reply:
Yes, deposits can be front-run which is why they need to be wrapped with the vault value checker.

Reply:
I've created PR https://github.com/OriginProtocol/origin-dollar/pull/2464 to add front-running protect on deposits